syntax="proto3";

service RaftProtocol {
  // RequestVote for raft leader node election.
  rpc RequestVote(VoteRequest) returns (VoteResponse) {}

  // AppendEntries for leader raft node to override/append its log entries into followers.
  rpc AppendEntries(AERequest) returns (AEResponse) {}
}

message VoteRequest {
  int64 term=1;
  string candidate_id=2;
  int64 last_log_index=3;
  int64 last_log_term=4;
}

message VoteResponse {
  bool received_vote=1;
  string error=2;
}

message WriteCommand {
  string key=1;
  string value=2;
}

message LogEntry {
  int64 log_term = 1;
  WriteCommand command = 2;
}

message AERequest {
  int64 term=1;
  string leader_id=2;
  int64 prev_log_index=3;
  int64 prev_log_term=4;

  repeated LogEntry entries = 5;
  
  int64 leader_commit_index=6;
  bool is_heart_beat = 7;
}

message AEResponse {
  int64 term = 1;
  bool is_success=2;
  string error=3;
}